core/drivers/crypto/caam/caam_ctrl.c

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2017-2019 NXP
+ *
+ * Brief   CAAM Global Controller.
+ */
+
+/* Global includes */
+#include <initcall.h>
+#include <kernel/panic.h>
+#include <tee_api_types.h>
+
+/* Local includes */
+#include <caam_common.h>
+#include <caam_hal_cfg.h>
+#include <caam_hal_clk.h>
+#include <caam_hal_ctrl.h>
+#ifdef CFG_CRYPTO_HASH_HW
+#include <caam_hash.h>
+#endif
+#include <caam_jr.h>
+#include <caam_pwr.h>
+#include <caam_rng.h>
+#include <caam_utils_mem.h>
+
+/*
+ * Crypto driver initialization function
+ *
+ * Returns:
+ * TEE_SUCCESS              Success
+ * TEE_ERROR_GENERIC        Generic Error (driver init failure)
+ * TEE_ERROR_NOT_SUPPORTED  Driver not supported
+ */
+static TEE_Result crypto_driver_init(void)
+{
+	TEE_Result retresult = TEE_ERROR_GENERIC;
+	enum CAAM_Status retstatus = CAAM_FAILURE;
+
+	struct jr_cfg jr_cfg = { 0 };
+
+	/* Enable the CAAM Clock */
+	hal_clk_enable(true);
+
+	retstatus = hal_cfg_get_conf(&jr_cfg);
+	if (retstatus != CAAM_NO_ERROR) {
+		retresult = TEE_ERROR_NOT_SUPPORTED;
+		goto exit_init;
+	}
+
+	/* Initialize the CAAM Controller */
+	hal_ctrl_init(jr_cfg.base);
+
+	/* Initialize the Job Ring to be used */
+	retstatus = caam_jr_init(&jr_cfg);
+	if (retstatus != CAAM_NO_ERROR) {
+		retresult = TEE_ERROR_GENERIC;
+		goto exit_init;
+	}
+
+	/* Initialize the RNG Module */
+	retstatus = caam_rng_init(jr_cfg.base);
+	if (retstatus != CAAM_NO_ERROR) {
+		retresult = TEE_ERROR_GENERIC;
+		goto exit_init;
+	}
+
+#ifdef CFG_CRYPTO_HASH_HW
+	/* Initialize the Hash Module */
+	retstatus = caam_hash_init(jr_cfg.base);
+	if (retstatus != CAAM_NO_ERROR) {
+		retresult = TEE_ERROR_GENERIC;
+		goto exit_init;
+	}
+#endif /* CFG_CRYPTO_HASH_HW */
+
+	/* Everything is OK, register the Power Management handler */
+	caam_pwr_init();
+
+	retresult = TEE_SUCCESS;
+
+exit_init:
+	/*
+	 * Configure Job Rings to NS World
+	 * If the Driver Crypto is not used (CFG_CRYPTO_DRIVER = n)
+	 * JR0 is freed to be Non-Secure
+	 */
+	if (jr_cfg.base)
+		hal_cfg_setup_nsjobring(jr_cfg.base);
+
+	if (retresult != TEE_SUCCESS) {
+		EMSG("CAAM Driver initialization (0x%x)", retresult);
+		panic();
+	}
+
+	return retresult;
+}
+
+driver_init(crypto_driver_init);
+
+/*
+ * Crypto driver late initialization function to complete CAAM operation
+ *
+ * Returns:
+ * TEE_SUCCESS        Success
+ */
+static TEE_Result init_caam_late(void)
+{
+	int ret = 0;
+
+	ret = caam_jr_complete();
```

Comment:
Why completing JR jobs during late init? Does this implicitly expect no other "late init" function is allowed to run JR jobs?

Reply:
This is to ensure that there is no pending jobs running in CAAM HW before leaving OPTEE boot. This is mandatory on certain platforms to continue the authentication boot processus.

Reply:
It's not clear when new jobs are allowed to be started. `caam_ctrl.c` may be used by other platforms not following your boot process, no?
(note that while i push this comment, i know nothing about such platforms. Feel free to discard my comment if you think the architecture/implementation is safe)

Reply:
@etienne-lms 
Well, I would not explain how to CAAM HW is working but CAAM is like a coprocessor that is running jobs while there is one present in the input queue. When a job is complete CAAM pushes information in the output queue and we have to take status completion from this queue.
As the CAAM is running independently and because of the different boot process on NXP devices, we have to ensure that there is no running job in the CAAM to continue.

caam_ctrl.c is specific to the CAAM HW and the driver itself. I don't see the interest for another platform to re-use it if it's HW doesn't work like ours.

I really appreciate the review you are doing, that help me to fix hidden issues. For sure, the architecture/implementation is safe in the sense where it's working since years on different OPTEE version and other internal libraries. We can still have bug because of the integration of the driver in OPTEE but in terms of architecture I think we are clean now.